### ISSUE TITLE
Do not dynamically regenerate rotations on view

### ISSUE THREAD

**Issue by mstenta** [farmOS/farm_grazing#20]
Currently the Rotations page is dynamically recalculating rotations whenever the page is loaded, regardless of what the actual saved rotation (and log) record has stored in the database. This can lead to issues when things change elsewhere in the system.

For example, if a paddock is removed from the plan, all rotations and logs that were associated with that paddock are removed (see https://github.com/mstenta/farm_grazing/issues/12).  However, it does not recalculate the rotations for the plan at that time, it just presents a message to the user telling them to review and update it themselves. When they go back to the Rotations page, the page dynamically recalculates the rotations and displays the new values... but it does NOT save those values until you click "Save" at the bottom.  So what you are looking at in the form is NOT what is actually stored in the database, and most importantly is NOT aligned with what the actual dates of the movement logs are.  So your todo list (logs) still shows the old rotation dates, but the Rotations page shows the recalculated dates.  I tagged this as a "bug" because of this.

You can fix this by clicking the "Save" button at the bottom of the page, but that assumes you understand that there is a mismatch in the first place.  We should not assume this, and we should not require the user to click the "Save" button if they are not actually making changes in the form.

I see two possible solutions:

1. Recalculate the rotations when paddocks are removed (as well as any other scenario that would cause this same problem).
2. Do not dynamically recalculate rotations when the Rotations page is loaded.

I'm not a fan of #1 because it is just a band-aid to the symptoms, but if we miss a scenario the bug will still exist.

I think the right solution to this is #2: do not dynamically recalculate rotations when the Rotations page is loaded.  Only recalculate when the form is submitted.

It should show the old saved values, even if they no longer make sense.  It should be the responsibility of the user to review and recalculate themselves.  By involving the user in the process, they will be forced to understand and consider what their changes (ie: removing a paddock from the plan) mean for the rest of the plan.  I think that's a good thing.

It also simplifies the code for building the form.  Rather than having all the same logic for calculating rotations, it only needs to display records from the database.  All calculations can be focused in the submit function.

(Note that this approach should still work with a fully-AJAX version of the rotation form, because it will still be calling the submit function when things change.)

**Comment by woodbri** [comment]
OK, so you are suggesting that we pull the data from the logs and use that to populate the table. 

- How do we do that when we are building a new plan that does not have any logs yet?
- This implies that we do away with the rotations table or that it is only used as a temp table for collecting data need for the calculations

I'm not sure I see how the workflow will be for the initial case planning case or how we should separate planning from the rendering.

**Comment by mstenta** [comment]
> so you are suggesting that we pull the data from the logs and use that to populate the table.

Nope - I'm thinking we just move all the calculation logic to the submit function, rather than the form builder function (which then only needs to show existing rotations data).

So when you initially select the herds and paddocks to add rotations for, you do all the necessary calculations in the submit function, save the rotation record to {farm_grazing_rotations} and create a log linked to it. And when you drag/edit the rotations, you save them (and the logs) on submit.

It wouldn't be all that different from what we have already I think. The only distinction is that it would not do any calculations in the function that builds and displays the form. Only in the submit functions, which save the data to the database. This just always ensures that what you see on the page is what is stored in the database, so you don't end up with situations where there is a disconnect between what you see and what is saved.

**Comment by woodbri** [comment]
I'm not quite sure I understand how your proposal works. I was thinking that the the rotations page should ALWAYS reflect the current state of the system, period. To that end, we make it AJAX based and there is not [Save] bottom or it there is it is used like a git commit to save a named branch (aka: a snapshot of the current state)

If someone changes something elsewhere, like deleting a paddock or changes animal inventory, etc. Then we either those functions call an updater or the render function updates any changes.  So I see one function, the computes the data needed to render and it makes updates as needed if something changes. This function can be called by and other functions that make changes, or by the AJAX handler and/or by the render function. So all computation and updates are done in a single place. If the delete paddock function forgets to call it, it still will get called when rotations tab is rendered.

**Comment by woodbri** [comment]
So to go back to your example, someone, other than the one viewing, deletes a paddock. 
What do a display when a render the form?
How does the user know the something is wrong with the data, ie: someone else deleted the paddock?

I think that the computation bit needs to detect that something has change, ie: compute the form based on the current data, and compare that to the logs and/or DB and then make changes to update the logs and DB so they are back in sync and alert the viewer that things have changed.

**Comment by mstenta** [comment]
> I'm not quite sure I understand how your proposal works.

Maybe I made it sound more complicated than it is. Also, some things have changed since the original description was written, so it may require a bit of an update.

For one, it is now impossible to delete records that a plan is using (ie: paddocks, logs, herds, animals). So those are no longer an issue.

There are still some considerations for things changing elsewhere in the system, aside from record deletions, but that wasn't my intention to solve all of those in this issue.

> I was thinking that the the rotations page should ALWAYS reflect the current state of the system, period.

Exactly.  That was the original intention with this issue.  That is NOT how the rotations page is working currently.

Every time the function `farm_grazing_plan_rotations_form()` runs (which is responsible for building the rotations page), it is dynamically recalculating all the rotations, and ignoring actual values stored in the database.  If you click save, then it saves them to the database, but if you don't, then what you see is NOT what is actually stored.

In my opinion, the function `farm_grazing_plan_rotations_form()` should only be building the form based on actual stored values. It should NOT be doing any actual calculations. The calculations should only be done if/when the user submits the form.

> If someone changes something elsewhere, like deleting a paddock or changes animal inventory, etc. Then we either those functions call an updater or the render function updates any changes. So I see one function, the computes the data needed to render and it makes updates as needed if something changes. This function can be called by and other functions that make changes, or by the AJAX handler and/or by the render function. So all computation and updates are done in a single place. If the delete paddock function forgets to call it, it still will get called when rotations tab is rendered.

Yea, I agree we should do this - consolidate all the updating into a single function that can be called from other places (where it makes sense).  But I consider that a separate issue.  This one is purely about fixing the potential disconnect that currently exists between what you see in the rotations page and what is currently stored in the database.

> So to go back to your example, someone, other than the one viewing, deletes a paddock.
What do a display when a render the form?
How does the user know the something is wrong with the data, ie: someone else deleted the paddock?

We don't have to worry about this anymore, because the system will prevent the paddock from being deleted because it is used in the plan.  This was done as part of Phase 1 (addition of the farm_constrain module that allows modules to define constraints/references between various entities that should prevent the ability to delete them).

Hopefully that clarifies my proposal a bit?  Might be easier to talk it through in gchat...

**Comment by mstenta** [comment]
> So to go back to your example, someone, other than the one viewing, deletes a paddock.

Oops - so there are actually two separate scenarios... I only addressed one, and didn't really answer your question I think.

1. User tries to DELETE a paddock that is linked to a plan.
2. User simply UNCHECKS a paddock within the plan, so that it is no longer an available option in the plan.

1 cannot happen anymore, because of the farm_constrain module I mentioned.

2 can still happen, however. I at least fixed part of it (the part that was creating orphan DB entries) in #12 - but that doesn't necessarily fix the usability side of it. Nothing is shown to the user to let them know that they need to adjust their plan if the paddock was removed. So we should think about that... but I still consider that a separate issue from this one.

**Comment by mstenta** [comment]
So to summarize the goal of this issue very simply:

Move all rotation calculation logic OUT of the function `farm_grazing_plan_rotations_form()` and INTO the function `farm_grazing_plan_rotations_form_submit()`.  So `farm_grazing_plan_rotations_form()` only shows what is ACTUALLY stored in the database.

**Comment by woodbri** [comment]
Ok, sounds like a go as we need the first piece done before we can test breaking it and resolving those potential issues

**Comment by mstenta** [comment]
Ok great - I'll see if I can knock this one out, and then we can create followup issues for next steps...

**Comment by mstenta** [comment]
Made some progress on this...

The form builder function only displays what is currently stored in the database now. I moved all the calculation logic into other functions...

I created a new function called `farm_grazing_recalculate_rotations($plan_id, $herd_id, $rotations = array(), $reset_durations = 0);`, which is a central place for all the logic around recalculating the start times and durations of each rotation for a given herd in a plan.  It takes a `$plan_id` and a `$herd_id` as required arguments, optionally accepts an array of rotations to work with (if not provided, they will be loaded from the database), and has an optional parameter for resetting rotation durations (0 = no change, 1 = minimum grazing days, 2 = average, 3 = maximum).

I changed the "Save changes" button at the bottom of the page to say "Recalculate", and moved the "Reset grazing days" option down to the same area, and wrapped both in a collapsed fieldset.

So now, the rotations page form submit function just calls out to the new `farm_grazing_recalculate_rotations()` function. If the user rearranged the rotations, and/or changed any of the "grazing days" field values, those changes will be made to the rotations and then passed into the recalculation function.

One thing this allowed me to do is remove the `weight` column from the `{farm_grazing_rotations}` table. The weight is used only in the form, but ultimately, it is just the `start_date` of each rotation that determines the order in which they take place, so storing an additional `weight` wasn't necessary.

Here is a quick summary of what the form submit function does:

- Iterate through the herds in the form
    - Iterate through the rotations for each herd
        - Keep track of any changes the user made to the grazing days and sort order fields
    - Reorder the rotations based on the new sort order
    - Send the rotations to the recalculation function for final calculations and saving to the database

Here's a quick summary of what the recalculation function does:

- Load rotations from the database (if they are not provided as an argument)
- Build grazing days data for the plan (if one of the "Reset grazing days" options was picked)
- Iterate through the rotations
    - Load the movement log (throw an error if there isn't one)
    - Update the rotation and log timestamps
    - Recalculate the duration, if desired
    - Save the rotation and log to the database
    - Calculate the next rotation's timestamp based on the duration

We can use this recalculation function elsewhere moving forward, and be confident that the same logic will always be applied. I also used the recalculation function when rotations are deleted, which simplified some of that code. I will probably also use it here: https://github.com/mstenta/farm_grazing/issues/16

**Comment by mstenta** [comment]
Pushed to a new branch "rotations_form": https://github.com/mstenta/farm_grazing/compare/rotations_form

**Comment by woodbri** [comment]
Regarding the weight column, it is used to reorder the paddocks when there is no JS available. So what was happening, is that you reordered the rows via drag&drop or using the weights, then when you saved it the start dates were recomputed using the last rec's this_row_start_date =  last_row_start_date+last_row_duration

**Comment by mstenta** [comment]
This is done! We can create new issues for other followup tasks if we need to.
